components/collections/codepointtrie_builder/src/native.rs

```diff
@@ -0,0 +1,152 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+use crate::CodePointTrieBuilder;
+use crate::CodePointTrieBuilderData;
+
+use icu_collections::codepointtrie::TrieType;
+use icu_collections::codepointtrie::TrieValue;
+use icu_collections::codepointtrie::{CodePointTrie, CodePointTrieHeader};
+
+use zerovec::ZeroVec;
+
+use core::{mem, slice};
+
+enum UMutableCPTrie {}
+
+#[repr(C)]
+#[allow(non_snake_case)]
+pub struct UCPTrie {
+    pub index: *const u16,
+    pub data: UCPTrieData,
+    pub indexLength: i32,
+    pub dataLength: i32,
+    pub highStart: u32,
+    pub shifted12HighStart: u16,
+    pub type_: i8,
+    pub valueWidth: i8,
+    pub reserved32: u32,
+    pub reserved16: u16,
+    pub index3NullOffset: u16,
+    pub dataNullOffset: i32,
+    pub nullValue: u32,
+}
+
+#[repr(C)]
+pub union UCPTrieData {
+    pub ptr0: *const ::std::os::raw::c_void,
+    pub ptr16: *const u16,
+    pub ptr32: *const u32,
+    pub ptr8: *const u8,
+}
+
+extern "C" {
+    fn umutablecptrie_open_72(
+        initial_value: u32,
+        error_value: u32,
+        error_code: &mut u32,
+    ) -> *const UMutableCPTrie;
+    fn umutablecptrie_set_72(
+        trie: *const UMutableCPTrie,
+        cp: u32,
+        value: u32,
+        error_code: &mut u32,
+    ) -> *const UMutableCPTrie;
+    fn umutablecptrie_buildImmutable_72(
+        trie: *const UMutableCPTrie,
+        trie_type: u32,
+        width: u32,
+        error_code: &mut u32,
+    ) -> *const UCPTrie;
+
+    fn ucptrie_close_72(trie: *const UCPTrie);
+    fn umutablecptrie_close_72(builder: *const UMutableCPTrie);
+}
+
+pub(crate) fn run_native<T>(cpt_builder: &CodePointTrieBuilder<T>) -> CodePointTrie<'static, T>
+where
+    T: TrieValue + Into<u32>,
+{
+    let mut error = 0;
+    let builder = unsafe {
+        umutablecptrie_open_72(
+            cpt_builder.default_value.into(),
+            cpt_builder.error_value.into(),
+            &mut error,
+        )
+    };
+
+    if error != 0 {
+        panic!("cpt builder returned error code {}", error);
```

Comment:
Thought: Not sure if we want to be panicky or return a Result. We're pointing at what we believe to be a real ICU library file, so maybe panic is fine here.

Reply:
Yeah, I feel the same.

Also making this return a result requires RAIIing all of the C types, and that's a mess.